docs/DeveloperGuide.adoc

```diff
@@ -218,29 +228,431 @@ image::CommitActivityDiagram.png[]
+* `TagCountManager` maps `Tag` to `Integer`, where `Integer` is the number of `Expense` under each `Tag`. It allows `Tag` objects that are not tagged with any `Expense` to be tracked and removed.
+
+Operations include:
+
+* `UniqueTagList#retrieveTags(List<String>)` -- Retrieves corresponding tags from `UniqueTagList` based on the list of tag names.
+* `UniqueTagList#removeAll(List<Tag>)` -- Removes tags given in the list from the `UniqueTagList`.
+* `UniqueTagList#getTagNames()` -- Returns a list of tag names.
+* `TagCountManager#incrementAllCount(Set<Tag>)` -- Increment the number of expenses mapped to tags given in the set by 1.
+* `TagCountManager#decreaseAllCount(Set<Tag>)` -- Decrease number of expenses mapped to tags given in the set by 1.
+* `TagCountManager#removeZeroCount()` -- Removes all mappings where number of expense is equal to 0.
+
+These operations are exposed in the `Model` interface as:
+
+* `Model#retrieveTags(List<String>)`
+* `Model#incrementCount(Set<Tag>)`
+* `Model#decreaseCount(Set<Tag>)` -- `TagCountManager#removeZeroCount()` and `UniqueTagList#removeAll(List<Tag>)` is called in this method as well to remove tags that does not have any expense tagged to it from `UniqueTagList` and `TagCountManager`.
+* `Model#getTagNames()`
+
+Given below is an example usage scenario and how the adding tag mechanism behaves at each step.
+
+Step 1.
+The user launches the application. The `Model` is initialized with saved data. All tags are loaded into `UniqueTagList` and `TagCountManager`.
+
+Step 2.
+User enters the command `tag add 1 t/test t/test2` to add tags to the 1st `Expense` in Billboard. `BillboardParser` parses the command, creating a `TagCommandParser`.
+The `TagCommandParser` then parses `add 1 t/test t/test2` and creates an `AddTagCommandParser`. Subsequently, the `AddTagCommandParser` parses `1 t/test1 t/test2` into `Index` 1 and
+a list of `String` consisting of `test1` and `test2`. They are then used to create `AddTagCommand` which is returned to `LogicManager`.
+
+Step 3.
+`LogicManager` executes the `AddTagCommand`. During execution, `AddTagCommand` calls `Model#retrieveTags(Set<String>)` on the list consisting of `test1` and `test2` which returns a set of `Tag`.
+`Model#incrementCount(Set<Tag>)` is then called on the set of `Tag`. The 1st `Expense` is edited and updated in the `Model` using `Model#setExpense(Expense, Expense)` and is then shown on the UI.
+
+[NOTE]
+Duplicate tags in an `Expense` is not allowed.
+If the user tries to add existing `Tag` to an `Expense`, it will not be added, and `Model#incrementCount` will not be called.
+If the user tries to add duplicate `Tag`, i.e enter 2 of the same `Tag`, it will only be added once and number of expense under that `Tag` will only increment by 1.
+
+The following sequence diagram shows how the adding tag operation works.
+
+image::AddTagSequenceDiagram.png[]
```

Comment:
`retrieveTag`'s return object is not defined in this sequence diagram.

Reply:
also the lines should stop after the cross, though we are aware of the technical difficulty of doing so.